src/main/java/org/betonquest/betonquest/compatibility/citizens/CitizensVariable.java

```diff
@@ -7,65 +7,78 @@
 import org.betonquest.betonquest.api.Variable;
 import org.betonquest.betonquest.api.profiles.Profile;
 import org.betonquest.betonquest.exceptions.InstructionParseException;
-import org.bukkit.Location;
+import org.betonquest.betonquest.variables.LocationVariable;
 
+import java.util.Arrays;
 import java.util.Locale;
 
 /**
  * Provides information about a citizen npc.
  * <p>
  * Format:
- * {@code %citizen.<id>.<type>%}
+ * {@code %citizen.<id>.<argument>.<mode>.<precision>%}
  * <p>
- * Types:
+ * Arguments:
  * * name - Return citizen name
  * * full_name - Full Citizen name
- * * location  - Return citizen location. x;y;z;world;yaw;pitch
+ * * location - Return citizen location, defaults to ulfLong
+ * Modes:
+ * * xyz - The x, y and z location of the npc, separated by spaces
+ * * x - The x location of the npc
+ * * y - The y location of the npc
+ * * z - The z location of the npc
+ * * world - The world location of the npc
+ * * yaw - The yaw of the npc
+ * * pitch - The pitch of the npc
+ * * ulfShort - The location of the npc in the form x;y;z;world
+ * * ulfLong - The location of the npc in the form x;y;z;world;yaw;pitch
+ * Precision is decimals of precision desired, defaults to 0.
  */
 @SuppressWarnings("PMD.CommentRequired")
 public class CitizensVariable extends Variable {
     private final int npcId;
 
-    private final TYPE key;
+    private final ARGUMENT key;
 
+    private final LocationVariable location;
+
+    @SuppressWarnings({"PMD.CommentRequired", "PMD.AvoidLiteralsInIfCondition"})
     public CitizensVariable(final Instruction instruction) throws InstructionParseException {
         super(instruction);
 
-        npcId = instruction.getInt();
         try {
-            key = TYPE.valueOf(instruction.next().toUpperCase(Locale.ROOT));
-        } catch (final IllegalArgumentException e) {
-            throw new InstructionParseException("Invalid Type: " + instruction.current(), e);
+            final String[] splitInstruction = instruction.getInstruction().split("\\.");
+            if (splitInstruction.length < 3) {
+                throw new InstructionParseException("Not enough arguments");
+            } else {
+                npcId = Integer.parseInt(splitInstruction[1]);
+                key = ARGUMENT.valueOf(splitInstruction[2].toUpperCase(Locale.ROOT));
+            }
+
+            final String newInstruction = String.join(".", Arrays.copyOfRange(splitInstruction, 2, splitInstruction.length));
+            location = new LocationVariable(new Instruction(instruction.getPackage(), null, newInstruction));
+        } catch (final IllegalArgumentException exception) {
```

Comment:
I think the illegal argument exception is only thrown by the call `ARGUMENT.valueOf(splitInstruction[2].toUpperCase(Locale.ROOT));` and therefore it should only surround the related code and should give an better message what the issue is, like the old code

Reply:
resolved